### PR TITLE
[Improvement][C++] Use inherit to implement EdgesCollection

### PR DESCRIPTION
**Pull request by acezen**
## Proposed changes
- replace the `variant` of EdgesCollection with inheritance to implement polymorphism of different adj list type edges collection.
- Add cpp build process to JAVA CMakeList to make JAVA SDK base on a specific version of cpp SDK, not compile with the latest commit. That would help the breaking change in C++ SDK not to infect JAVA SDK imediately.

## Types of changes

What types of changes does your code introduce to GraphAr?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/alibaba/GraphAr/blob/main/CONTRIBUTING.rst) doc
- [x] I have signed the CLA
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by Thespica** [comment]
I think the best way to adapt this change is getting EdgesCollection from factory function directory.

For details, you need: 
- Remove 4 supper classes of EdgesCollection;
- Remove `create` function in EdgesCollection;
- Write a new factory function in `util/GrapharStaticFunctions.java` to create EdgesCollection, which maps `ConstructEdgesCollection` in C++;
- Add annotations for EdgesCollection:
  - For class: `@FFIGen` `@FFITypeAlias` `@CXXHead`
  - For return objects: `@CXXReference` or `@CXXValue`
- Modify related usage in test.

If there are new problems, please reach out to me.

File: cpp/examples/bfs_father_example.cc [hunk 1]
```diff
@@ -36,8 +36,8 @@ int main(int argc, char* argv[]) {
   auto maybe_vertices =
       GAR_NAMESPACE::ConstructVerticesCollection(graph_info, label);
   ASSERT(maybe_vertices.status().ok());
-  auto& vertices = maybe_vertices.value();
-  int num_vertices = vertices.size();
+  auto vertices = maybe_vertices.value();
```
Comment (lixueclaire) [inline cpp/examples/bfs_father_example.cc:39]: Maybe use `auto&` consistently is better?

Comment (acezen) [reply cpp/examples/bfs_father_example.cc]: Fixed.

**Comment by acezen** [comment]
@lixueclaire any comments?

**Review by lixueclaire — APPROVED** [review]
LGTM

**Review by Thespica — CHANGES_REQUESTED** [review]

File: java/CMakeLists.txt [hunk 2]
```diff
@@ -19,10 +20,12 @@ include_directories("src/main/native")
 include_directories("src/test/native")
 
 find_package(Arrow REQUIRED)
-find_package(gar REQUIRED)
+include(graphar-cpp)
+build_graphar_cpp()
```
[the remark this answers is not in the document]
Comment (acezen) [reply java/CMakeLists.txt:24]: The C++ library is actually just build for the first time, so i think it's ok for that.

**Pull request merged**